--- src/makefile.unix~	2016-12-05 00:11:44.000000000 +0200
+++ src/makefile.unix	2017-01-24 12:10:11.319657735 +0200
@@ -11,6 +11,8 @@
 # :=0 --> Disable IPv6 support
 USE_IPV6:=1
 
+USE_SYSTEM_LEVELDB:=0
+
 LINK:=$(CXX)
 
 DEFS=-DBOOST_SPIRIT_THREADSAFE -D_FILE_OFFSET_BITS=64
@@ -188,7 +190,6 @@
      obj/transports.o \
      obj/util_codedigest.o \
      obj/joincoin.o \
-    leveldb/libleveldb.a \
     obj/alert.o \
     obj/version.o \
     obj/checkpoints.o \
@@ -250,11 +251,16 @@
 # LevelDB support
 #
 MAKEOVERRIDES =
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
 LIBS += $(CURDIR)/leveldb/libleveldb.a $(CURDIR)/leveldb/libmemenv.a
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
+OBJS += leveldb/libleveldb.a
 leveldb/libleveldb.a:
 	@echo "Building LevelDB ..." && cd leveldb && $(MAKE) CC=$(CC) CXX=$(CXX) OPT="$(xCXXFLAGS)" libleveldb.a libmemenv.a && cd ..
+else
+     LIBS += -lleveldb -lmemenv
+endif
 
 # auto-generated dependencies:
 #-include obj/*.P
--- src/leveldb.cpp~	2016-12-05 00:11:44.000000000 +0200
+++ src/leveldb.cpp	2017-01-24 12:10:53.812331874 +0200
@@ -8,7 +8,7 @@
 #include <leveldb/env.h>
 #include <leveldb/cache.h>
 #include <leveldb/filter_policy.h>
-#include <memenv/memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 #include <boost/filesystem.hpp>
 
