--- src/makefile.unix~	2015-11-29 16:58:41.000000000 +0200
+++ src/makefile.unix	2015-11-29 17:09:22.321646078 +0200
@@ -10,6 +10,10 @@
 # :=1 --> Enable IPv6 support
 # :=0 --> Disable IPv6 support
 USE_IPV6:=1
+ 
+# :=1 --> Use system leveldb
+# :=0 --> Use embedded leveldb
+USE_SYSTEM_LEVELDB:=0
 
 LINK:=$(CXX)
 
@@ -113,7 +117,6 @@
 # anoncoind depends on them. This is inelegant because they are redundant when listed
 # in $OBJS (NOTE: compilation fails if they are not in $LIBS, even if present here).
 OBJS= \
-    leveldb/libleveldb.a \
     ../i2psam/libi2psam.a \
     obj/alert.o \
     obj/version.o \
@@ -150,6 +153,9 @@
     obj/txdb.o \
     obj/miner.o
 
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
+     OBJS += leveldb/libleveldb.a
+endif
 
 ifdef USE_SSE2
 DEFS += -DUSE_SSE2
@@ -166,11 +172,15 @@
 # LevelDB support
 #
 MAKEOVERRIDES =
-LIBS += $(CURDIR)/leveldb/libleveldb.a $(CURDIR)/leveldb/libmemenv.a
-DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
-DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
-leveldb/libleveldb.a:
-	@echo "Building LevelDB ..." && cd leveldb && $(MAKE) CC=$(CC) CXX=$(CXX) OPT="$(xCXXFLAGS)" libleveldb.a libmemenv.a && cd ..
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
+	LIBS += $(CURDIR)/leveldb/libleveldb.a $(CURDIR)/leveldb/libmemenv.a
+	DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
+	DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
+	leveldb/libleveldb.a:
+		@echo "Building LevelDB ..." && cd leveldb && $(MAKE) CC=$(CC) CXX=$(CXX) OPT="$(xCXXFLAGS)" libleveldb.a libmemenv.a && cd ..
+else
+	LIBS += -lleveldb -lmemenv
+endif
 
 #
 # I2P support
--- src/leveldb.cpp~	2015-11-03 11:13:20.000000000 +0200
+++ src/leveldb.cpp	2015-11-29 17:12:40.374651381 +0200
@@ -8,7 +8,7 @@
 #include <leveldb/env.h>
 #include <leveldb/cache.h>
 #include <leveldb/filter_policy.h>
-#include <memenv/memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 #include <boost/filesystem.hpp>
 
