--- configure.ac~	2014-10-08 11:21:17.000000000 +0300
+++ configure.ac	2015-11-02 17:03:52.304185594 +0200
@@ -378,10 +378,22 @@
  [ AC_MSG_RESULT(no)]
 )
 
+dnl Check for leveldb, only if explicitly requested
 LEVELDB_CPPFLAGS=
 LIBLEVELDB=
 LIBMEMENV=
-AM_CONDITIONAL([EMBEDDED_LEVELDB],[true])
+AC_ARG_WITH([system-leveldb],
+  [AS_HELP_STRING([--with-system-leveldb],
+  [Build with system LevelDB (default is no; DANGEROUS; NOT SUPPORTED)])],
+  [system_leveldb=$withval],
+  [system_leveldb=no]
+)
+if test x$system_leveldb != xno; then
+  LEVELDB_CPPFLAGS=
+  LIBLEVELDB=-lleveldb
+  LIBMEMENV=-lmemenv
+fi
+AM_CONDITIONAL([EMBEDDED_LEVELDB],[test x$system_leveldb = xno])
 AC_SUBST(LEVELDB_CPPFLAGS)
 AC_SUBST(LIBLEVELDB)
 AC_SUBST(LIBMEMENV)
--- src/leveldbwrapper.cpp~	2014-10-08 11:21:17.000000000 +0300
+++ src/leveldbwrapper.cpp	2015-11-02 17:05:18.666556817 +0200
@@ -10,7 +10,7 @@
 #include <leveldb/cache.h>
 #include <leveldb/env.h>
 #include <leveldb/filter_policy.h>
-#include <memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 void HandleError(const leveldb::Status &status) throw(leveldb_error) {
     if (status.ok())
--- src/dbwrapper.cpp~	2017-10-28 22:26:51.000000000 +0300
+++ src/dbwrapper.cpp	2017-11-22 21:47:22.603399852 +0200
@@ -12,7 +12,7 @@
 #include <leveldb/cache.h>
 #include <leveldb/env.h>
 #include <leveldb/filter_policy.h>
-#include <memenv.h>
+#include <leveldb/helpers/memenv.h>
 #include <stdint.h>
 
 static leveldb::Options GetOptions(size_t nCacheSize)
