--- configure.ac~	2017-03-01 22:31:53.000000000 +0200
+++ configure.ac	2017-03-12 15:05:50.737868568 +0200
@@ -541,7 +541,18 @@
 LEVELDB_CPPFLAGS=
 LIBLEVELDB=
 LIBMEMENV=
-AM_CONDITIONAL([EMBEDDED_LEVELDB],[true])
+AC_ARG_WITH([system-leveldb],
+  [AS_HELP_STRING([--with-system-leveldb],
+  [Build with system LevelDB (default is no; DANGEROUS; NOT SUPPORTED)])],
+  [system_leveldb=$withval],
+  [system_leveldb=no]
+)
+if test x$system_leveldb != xno; then
+  LEVELDB_CPPFLAGS=
+  LIBLEVELDB=-lleveldb
+  LIBMEMENV=-lmemenv
+fi
+AM_CONDITIONAL([EMBEDDED_LEVELDB],[test x$system_leveldb = xno])
 AC_SUBST(LEVELDB_CPPFLAGS)
 AC_SUBST(LIBLEVELDB)
 AC_SUBST(LIBMEMENV)
--- src/dbwrapper.cpp~	2017-03-01 22:31:53.000000000 +0200
+++ src/dbwrapper.cpp	2017-03-12 15:06:37.713867344 +0200
@@ -12,7 +12,7 @@
 #include <leveldb/cache.h>
 #include <leveldb/env.h>
 #include <leveldb/filter_policy.h>
-#include <memenv.h>
+#include <leveldb/helpers/memenv.h>
 #include <stdint.h>
 
 void HandleError(const leveldb::Status& status) throw(dbwrapper_error)
