--- CMakeLists.txt~	2015-09-24 16:04:17.301388657 +0300
+++ CMakeLists.txt	2015-09-24 16:04:39.886257806 +0300
@@ -447,7 +447,7 @@
 	add_subdirectory(secp256k1)
 endif ()
 
-add_subdirectory(libscrypt)
+# add_subdirectory(libscrypt)
 add_subdirectory(libdevcrypto)
 
 if (GENERAL)
@@ -471,6 +471,7 @@
 endif ()
 
 if (MINER OR TOOLS)
+        add_subdirectory(libevmcore)
 	add_subdirectory(ethminer)
 endif ()
 
