diff --git a/src/class_outgoingSynSender.py b/src/class_outgoingSynSender.py
index 91cd660..7e9cf57 100644
--- a/src/class_outgoingSynSender.py
+++ b/src/class_outgoingSynSender.py
@@ -216,12 +216,12 @@ class outgoingSynSender(threading.Thread, StoppableThread):
 
                 logger.debug(str(self) + ' connected to ' + str(peer) + ' during an outgoing attempt.')
             except socks.GeneralProxyError as err:
-                if err[0][0] in [7, 8, 9]:
+                if err.socket_err in [7, 8, 9]:
                     logger.error('Error communicating with proxy: %s', str(err))
                     queues.UISignalQueue.put((
                         'updateStatusBar',
                         tr._translate(
-                            "MainWindow", "Problem communicating with proxy: %1. Please check your network settings.").arg(str(err[0][1]))
+                            "MainWindow", "Problem communicating with proxy: %1. Please check your network settings.").arg(err.msg)
                         ))
                     self.stop.wait(1)
                     continue
@@ -246,20 +246,20 @@ class outgoingSynSender(threading.Thread, StoppableThread):
                 if deletedPeer:
                     str ('deleting ' + str(peer) + ' from knownnodes.knownNodes because it is more than 48 hours old and we could not connect to it.')
 
-            except socks.Socks5AuthError as err:
+            except socks.SOCKS5AuthError as err:
                 queues.UISignalQueue.put((
                     'updateStatusBar', tr._translate(
                     "MainWindow", "SOCKS5 Authentication problem: %1. Please check your SOCKS5 settings.").arg(str(err))))
-            except socks.Socks5Error as err:
-                if err[0][0] in [3, 4, 5, 6]:
+            except socks.SOCKS5Error as err:
+                if err.socket_err in [3, 4, 5, 6]:
                     # this is a more bening "error": host unreachable, network unreachable, connection refused, TTL expired
                     logger.debug('SOCKS5 error: %s', str(err))
                 else:
                     logger.error('SOCKS5 error: %s', str(err))
-                if err[0][0] == 4 or err[0][0] == 2:
+                if err.socket_err in [4, 2]:
                     state.networkProtocolAvailability[protocol.networkType(peer.host)] = False
-            except socks.Socks4Error as err:
-                logger.error('Socks4Error: ' + str(err))
+            except socks.SOCKS4Error as err:
+                logger.error('SOCKS4 error: ' + str(err))
             except socket.error as err:
                 if BMConfigParser().get('bitmessagesettings', 'socksproxytype')[0:5] == 'SOCKS':
                     logger.error('Bitmessage MIGHT be having trouble connecting to the SOCKS server. ' + str(err))
