diff --git a/src/bitmessageqt/__init__.py b/src/bitmessageqt/__init__.py
index 4cda426..d37f79a 100644
--- a/src/bitmessageqt/__init__.py
+++ b/src/bitmessageqt/__init__.py
@@ -2541,7 +2541,7 @@ class MyForm(settingsmixin.SMainWindow):
             tableWidget.item(i, 2).setUnread(False)
             tableWidget.item(i, 3).setFont(font)
             # sqlite default limit, unfortunately getting/setting isn't exposed to python
-            if i % 999 == 999: 
+            if i % 999 == 0:
                 markread += partialUpdate(self.getCurrentFolder(), msgids)
                 msgids = []
 
