diff --git a/src/protocol.py b/src/protocol.py
index 9a0c6a11..eaba973c 100644
--- a/src/protocol.py
+++ b/src/protocol.py
@@ -262,7 +262,8 @@ def assembleVersionMessage(remoteHost, remotePort, participatingStreams, server=
         payload += encodeHost('127.0.0.1')
         payload += pack('>H', 8444)
     else:
-        payload += encodeHost(remoteHost)
+        # 127.0.0.1 for onion v3 services
+        payload += encodeHost(remoteHost if len(remoteHost) <= 22 else '127.0.0.1')
         payload += pack('>H', remotePort)  # remote IPv6 and port
 
     # bitflags of the services I offer.
