diff --git a/setup.py b/setup.py
index 422524bd..26960c24 100644
--- a/setup.py
+++ b/setup.py
@@ -77,7 +77,8 @@ if __name__ == "__main__":
 
     try:
         import msgpack
-        installRequires.append("msgpack-python")
+        installRequires.append(
+            "msgpack-python" if msgpack.version[:2] == (0, 4) else "msgpack")
     except ImportError:
         try:
             import umsgpack
