diff --git a/src/knownnodes.py b/src/knownnodes.py
index 5755efc0..5ffd65d1 100644
--- a/src/knownnodes.py
+++ b/src/knownnodes.py
@@ -37,6 +37,7 @@ def json_serialize_knownnodes(output):
     _serialized = []
     for stream, peers in knownNodes.iteritems():
         for peer, info in peers.iteritems():
+            info.update(rating=round(info.get('rating', 0), 2))
             _serialized.append({
                 'stream': stream, 'peer': peer._asdict(), 'info': info
             })
