--- src/net.cpp~	2018-03-02 17:04:18.000000000 +0200
+++ src/net.cpp	2018-03-02 17:06:43.241667987 +0200
@@ -1143,7 +1143,7 @@
 // The first name is used as information source for addrman.
 // The second name should resolve to a list of seed addresses.
 static const char *strDNSSeed[][2] = {
-    {"18.220.243.249", "18.220.205.65"}
+//    {"18.220.243.249", "18.220.205.65"}
 };
 
 void ThreadDNSAddressSeed(void* parg)
