--- src/net.cpp~	2019-12-11 15:21:55.000000000 +0200
+++ src/net.cpp	2019-12-11 15:55:58.448962954 +0200
@@ -19,10 +19,10 @@
 #endif
 
 #ifdef USE_UPNP
-#include <miniupnpc-2.1/miniupnpc.h>
-#include <miniupnpc-2.1/miniwget.h>
-#include <miniupnpc-2.1/upnpcommands.h>
-#include <miniupnpc-2.1/upnperrors.h>
+#include <miniupnpc/miniupnpc.h>
+#include <miniupnpc/miniwget.h>
+#include <miniupnpc/upnpcommands.h>
+#include <miniupnpc/upnperrors.h>
 #endif
 
 #include <boost/filesystem.hpp>
