commit 2815a0e9bdfba7f6bca8960196a5b9e1509e04a2
Author: Dmitri Bogomolov <4glitch@gmail.com>
Date:   Mon Oct 1 16:30:31 2018 +0300

    Fixed typo in dogecoin_find_bdb51.m4

diff --git a/build-aux/m4/dogecoin_find_bdb51.m4 b/build-aux/m4/dogecoin_find_bdb51.m4
index b17c7db6e..b33491090 100644
--- a/build-aux/m4/dogecoin_find_bdb51.m4
+++ b/build-aux/m4/dogecoin_find_bdb51.m4
@@ -47,7 +47,7 @@ AC_DEFUN([BITCOIN_FIND_BDB51],[
       AC_MSG_ERROR([Found Berkeley DB other than 5.1, required for portable wallets (--with-incompatible-bdb to ignore or --disable-wallet to disable wallet functionality)])
     ])
   else
-    BITCOIN_SUBDIR_TO_INCLUDE(BDB_CPPFLAGS,[${bdb48path}],db_cxx)
+    BITCOIN_SUBDIR_TO_INCLUDE(BDB_CPPFLAGS,[${bdb51path}],db_cxx)
     bdbpath="${bdb51path}"
   fi
   AC_SUBST(BDB_CPPFLAGS)
