--- src/txdb-leveldb.cpp~	2015-04-14 23:54:43.000000000 +0300
+++ src/txdb-leveldb.cpp	2016-11-21 17:00:16.547181663 +0200
@@ -12,7 +12,7 @@
 #include <leveldb/env.h>
 #include <leveldb/cache.h>
 #include <leveldb/filter_policy.h>
-#include <memenv/memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 #include "kernel.h"
 #include "checkpoints.h"

--- src/makefile.unix~	2016-11-21 17:28:52.000000000 +0200
+++ src/makefile.unix	2016-11-21 17:30:49.652598706 +0200
@@ -3,6 +3,7 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 USE_UPNP:=0
+USE_SYSTEM_LEVELDB:=0
 
 LINK:=$(CXX)
 ARCH:=$(system lscpu | head -n 1 | awk '{print $2}')
@@ -146,13 +147,18 @@
 
 all: bitbayd
 
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
 LIBS += $(CURDIR)/leveldb/libleveldb.a $(CURDIR)/leveldb/libmemenv.a
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
-OBJS += obj/txdb-leveldb.o
 leveldb/libleveldb.a:
 	@echo "Building LevelDB ..."; cd leveldb; make libleveldb.a libmemenv.a; cd ..;
 obj/txdb-leveldb.o: leveldb/libleveldb.a
+else
+LIBS += -lleveldb -lmemenv
+endif
+
+OBJS += obj/txdb-leveldb.o
 
 # auto-generated dependencies:
 -include obj/*.P
