--- src/txdb-leveldb.cpp~	2018-08-16 00:44:28.000000000 +0300
+++ src/txdb-leveldb.cpp	2018-08-23 13:51:33.503833418 +0300
@@ -12,7 +12,7 @@
 #include <leveldb/env.h>
 #include <leveldb/cache.h>
 #include <leveldb/filter_policy.h>
-#include <memenv/memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 #include "kernel.h"
 #include "checkpoints.h"
--- src/makefile.unix.orig	2018-08-16 00:44:28.000000000 +0300
+++ src/makefile.unix	2018-08-25 22:41:06.692579662 +0300
@@ -3,7 +3,7 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 # Build STATIC for this makefile (1=static, 0=not)
-STATIC=1
+# STATIC:=-
 # DEBUG=1
 
 INCLUDEPATHS= \
@@ -32,7 +32,8 @@
 
 USE_UPNP:=-
 USE_IPV6:=-
-
+USE_SYSTEM_LEVELDB:=0
+USE_SYSTEM_CRYPTOPP:=0
 
 LINK:=$(CXX)
 ARCH:=$(system lscpu | head -n 1 | awk '{print $2}')
@@ -345,19 +346,25 @@
 
 all: breakoutd
 
-
+ifeq (${USE_SYSTEM_CRYPTOPP}, 0)
 LIBS += $(CURDIR)/libcryptopp/libcryptopp.a
+else
+LIBS += -lcryptopp
+endif
+
 libcryptopp/libcryptopp.a:
 	@echo "Building libcryptopp ..."; cd libcryptopp; make libcryptopp.a
 
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
 LIBS += $(CURDIR)/leveldb/out-static/libleveldb.a $(CURDIR)/leveldb/out-static/libmemenv.a
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
-OBJS += obj/txdb-leveldb.o
 leveldb/out-static/libleveldb.a:
 	@echo "Building LevelDB ..."; cd leveldb; make staticlibs; cd ..;
 obj/txdb-leveldb.o: leveldb/out-static/libleveldb.a
-
+else
+LIBS += -lleveldb -lmemenv
+endif
 
 # auto-generated dependencies:
 -include obj/*.P
@@ -475,9 +482,8 @@
 obj/scrypt-arm.o: scrypt-arm.S
 	$(CXX) -c $(xCXXFLAGS) -MMD -o $@ $<
 
-breakoutd: leveldb/out-static/libleveldb.a libcryptopp/libcryptopp.a \
-              $(OBJS:obj/%=obj/%) obj/breakout.o
-	$(CXX) $(xCXXFLAGS) -o $@ $(LIBPATHS) $^ $(xLDFLAGS) $(LIBS)
+breakoutd: libcryptopp/libcryptopp.a $(OBJS:obj/%=obj/%) obj/breakout.o
+	$(LINK) $(xCXXFLAGS) -o $@ $(LIBPATHS) $^ $(xLDFLAGS) $(LIBS)
 
 clean:
 	-rm -f breakoutd
