--- configure.ac~	2018-03-10 19:21:14.000000000 +0200
+++ configure.ac	2018-04-20 16:11:23.887497118 +0300
@@ -386,7 +386,18 @@
 LEVELDB_CPPFLAGS=
 LIBLEVELDB=
 LIBMEMENV=
-AM_CONDITIONAL([EMBEDDED_LEVELDB],[true])
+AC_ARG_WITH([system-leveldb],
+  [AS_HELP_STRING([--with-system-leveldb],
+  [Build with system LevelDB (default is no; DANGEROUS; NOT SUPPORTED)])],
+  [system_leveldb=$withval],
+  [system_leveldb=no]
+)
+if test x$system_leveldb != xno; then
+  LEVELDB_CPPFLAGS=
+  LIBLEVELDB=-lleveldb
+  LIBMEMENV=-lmemenv
+fi
+AM_CONDITIONAL([EMBEDDED_LEVELDB],[test x$system_leveldb = xno])
 AC_SUBST(LEVELDB_CPPFLAGS)
 AC_SUBST(LIBLEVELDB)
 AC_SUBST(LIBMEMENV)
--- src/dbwrapper.cpp~	2019-06-12 14:08:46.000000000 +0300
+++ src/dbwrapper.cpp	2019-06-14 14:15:37.724541664 +0300
@@ -10,7 +10,7 @@
 #include <leveldb/cache.h>
 #include <leveldb/env.h>
 #include <leveldb/filter_policy.h>
-#include <memenv.h>
+#include <leveldb/helpers/memenv.h>
 #include <stdint.h>
 #include <algorithm>
 
