--- Makefile.am~	2016-11-28 19:59:56.000000000 +0200
+++ Makefile.am	2016-12-03 16:23:26.565596604 +0200
@@ -85,19 +85,6 @@
   algo/heavy/heavy.c \
   algo/heavy/bastion.c \
   algo/hmq1725.c \
-  algo/hodl/hodl.cpp \
-  algo/hodl/hodl-gate.c \
-  algo/hodl/hodl_arith_uint256.cpp \
-  algo/hodl/hodl_uint256.cpp \
-  algo/hodl/hash.cpp \
-  algo/hodl/hmac_sha512.cpp \
-  algo/hodl/sha256.cpp \
-  algo/hodl/sha512.cpp \
-  algo/hodl/utilstrencodings.cpp \
-  algo/hodl/hodl-wolf.c \
-  algo/hodl/aes.c \
-  algo/hodl/sha512_avx.c \
-  algo/hodl/sha512_avx2.c \
   algo/lbry.c \
   algo/luffa/luffa.c \
   algo/luffa/sse2/luffa_for_sse2.c \
--- algo-gate-api.c~	2016-11-28 19:59:56.000000000 +0200
+++ algo-gate-api.c	2016-12-03 16:29:13.305623409 +0200
@@ -164,7 +164,6 @@
      case ALGO_GROESTL:     register_groestl_algo    ( gate ); break;
      case ALGO_HEAVY:       register_heavy_algo      ( gate ); break;
      case ALGO_HMQ1725:     register_hmq1725_algo    ( gate ); break;
-     case ALGO_HODL:        register_hodl_algo       ( gate ); break;
      case ALGO_KECCAK:      register_keccak_algo     ( gate ); break;
      case ALGO_LBRY:        register_lbry_algo       ( gate ); break;
      case ALGO_LUFFA:       register_luffa_algo      ( gate ); break;
