--- src/makefile.unix~	2015-06-14 02:57:50.000000000 +0300
+++ src/makefile.unix	2018-11-29 11:05:48.614515294 +0200
@@ -11,6 +11,10 @@
 # :=0 --> Disable IPv6 support
 USE_IPV6:=1
 
+# :=1 --> Use system leveldb
+# :=0 --> Use embedded leveldb
+USE_SYSTEM_LEVELDB:=0
+
 LINK:=$(CXX)
 
 DEFS=-DBOOST_SPIRIT_THREADSAFE -D_FILE_OFFSET_BITS=64
@@ -111,7 +115,6 @@
 xLDFLAGS=$(LDHARDENING) $(LDFLAGS)
 
 OBJS= \
-    leveldb/libleveldb.a \
     obj/alert.o \
     obj/version.o \
     obj/checkpoints.o \
@@ -144,6 +147,9 @@
     obj/leveldb.o \
     obj/txdb.o
 
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
+     OBJS += leveldb/libleveldb.a
+endif
 
 ifdef USE_SSE2
 DEFS += -DUSE_SSE2
@@ -159,12 +165,16 @@
 #
 # LevelDB support
 #
+ifeq (${USE_SYSTEM_LEVELDB}, 0)
 MAKEOVERRIDES =
 LIBS += $(CURDIR)/leveldb/libleveldb.a $(CURDIR)/leveldb/libmemenv.a
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/include)
 DEFS += $(addprefix -I,$(CURDIR)/leveldb/helpers)
 leveldb/libleveldb.a:
 	@echo "Building LevelDB ..." && cd leveldb && $(MAKE) CC=$(CC) CXX=$(CXX) OPT="$(xCXXFLAGS)" libleveldb.a libmemenv.a && cd ..
+else
+LIBS += -lleveldb -lmemenv
+endif
 
 # auto-generated dependencies:
 -include obj/*.P
--- src/leveldb.cpp~	2015-06-14 02:57:50.000000000 +0300
+++ src/leveldb.cpp	2018-11-29 11:06:36.225969316 +0200
@@ -8,7 +8,7 @@
 #include <leveldb/env.h>
 #include <leveldb/cache.h>
 #include <leveldb/filter_policy.h>
-#include <memenv/memenv.h>
+#include <leveldb/helpers/memenv.h>
 
 #include <boost/filesystem.hpp>
 
