--- configure.ac~	2015-07-29 16:50:44.000000000 +0300
+++ configure.ac	2015-09-03 16:09:27.757737433 +0300
@@ -147,6 +147,9 @@
    LIBCURL_CFLAGS="$LIBCURL_CFLAGS -I$with_curl/include"
    LIBCURL_LDFLAGS="-L$with_curl/lib $LIBCURL_LDFLAGS"
    LIBCURL="-lcurl -lz"
+else
+LIBCURL_CHECK_CONFIG(, 7.15.2, ,
+  [AC_MSG_ERROR([Missing required libcurl >= 7.15.2])])
 fi
 
 # SSL install path (for mingw : --with-crypto=/usr/local/ssl)
