--- Makefile.am~	2021-03-30 17:36:09.109487410 +0300
+++ Makefile.am	2021-03-30 17:36:40.402059034 +0300
@@ -3,7 +3,7 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 ACLOCAL_AMFLAGS = -I build-aux/m4
-SUBDIRS = src/tor src
+SUBDIRS = src/secp256k1 src
 if ENABLE_MAN
 SUBDIRS += doc/man
 endif
@@ -169,7 +169,7 @@
 	$(MAKE) -C src $(@F)
 
 if USE_LCOV
-LCOV_FILTER_PATTERN=-p "/usr/include/" -p "src/leveldb/" -p "src/tor/" -p "src/bench/" -p "src/univalue" -p "src/crypto/ctaes" -p "src/secp256k1"
+LCOV_FILTER_PATTERN=-p "/usr/include/" -p "src/leveldb/" -p "src/bench/" -p "src/univalue" -p "src/crypto/ctaes" -p "src/secp256k1"
 
 baseline.info:
 	$(LCOV) -c -i -d $(abs_builddir)/src -o $@
--- configure.ac~	2021-03-30 18:34:22.068039269 +0300
+++ configure.ac	2021-03-30 18:35:46.377885126 +0300
@@ -1419,10 +1419,9 @@
   AC_CONFIG_SUBDIRS([src/univalue])
 fi
 
-AX_SUBDIRS_CONFIGURE([src/secp256k1], [[--disable-shared], [--with-pic], [--with-bignum=no], [--enable-module-recovery], [--enable-experimental], [--enable-module-ecdh]])
 
-ac_configure_args="${ac_configure_args} --disable-shared --with-pic --with-bignum=no --enable-module-recovery --disable-jni --disable-system-torrc --disable-systemd --disable-lzma --disable-zstd --disable-asciidoc"
-AC_CONFIG_SUBDIRS([src/tor])
+ac_configure_args="${ac_configure_args} --disable-shared --with-pic --with-bignum=no --enable-module-recovery --enable-experimental --enable-module-ecdh"
+AC_CONFIG_SUBDIRS([src/secp256k1])
 
 AC_OUTPUT
 
