--- src/net.cpp~	2018-09-26 11:37:38.000000000 +0300
+++ src/net.cpp	2018-09-26 11:38:47.041707175 +0300
@@ -61,7 +61,7 @@
 static CNode* pnodeSync = NULL;
 CAddress addrSeenByPeer(CService("0.0.0.0", 0), nLocalServices);
 uint64_t nLocalHostNonce = 0;
-array<int, THREAD_MAX> vnThreadsRunning;
+boost::array<int, THREAD_MAX> vnThreadsRunning;
 static std::vector<SOCKET> vhListenSocket;
 CAddrMan addrman;
 int nMaxConnections = 125;
